frontend/public/components/utils/resource-log.jsx

```diff
@@ -0,0 +1,201 @@
+import * as _ from 'lodash-es';
+import * as PropTypes from 'prop-types';
+import * as React from 'react';
+
+import { resourceURL, modelFor } from '../../module/k8s';
+import { SafetyFirst } from '../safety-first';
+import { LoadingInline, LogWindow, TogglePlay, LineBuffer, stream } from './';
+
+const dataHasFailureMsg = (data) => {
+  return _.includes(data, '"status": "Failure"');
+};
+
+const dataHasHTML = (data) => {
+  return _.includes(data, '<html') || _.includes(data, '<HTML');
+};
+
+// Component for the streaming controls
+const LogControls = ({status, toggleStreaming, label}) => {
+  return <div className="co-m-pane__top-controls">
+    { status === 'loading' && <span className="co-icon-space-l"><LoadingInline /></span> }
+    { ['streaming', 'paused', 'eof'].includes(status) && <TogglePlay active={status === 'streaming'} onClick={toggleStreaming} disabled={status === 'eof'}/> }
+    <span className="log-container-selector__text">
+      { status === 'loading' && <span>Loading log...</span> }
+      { status === 'streaming' && <span>Streaming logs from</span> }
+      { status === 'paused' && <span>Log stream is paused.</span> }
+      { status === 'eof' && <span>Log stream ended.</span> }
+    </span>
+    {label}
+  </div>;
+};
+
+export class ResourceLog extends SafetyFirst {
+  constructor(props) {
+    super(props);
+
+    this._buffer = new LineBuffer(this.props.bufferSize);
+    this._loadStarted = this._loadStarted.bind(this);
+    this._processData = this._processData.bind(this);
+    this._retry = _.throttle(this._retry, 5000);
+    this._toggleStreaming = this._toggleStreaming.bind(this);
+    this._updateStatus = this._updateStatus.bind(this);
+
+    this.state = {
+      url: '',
+      touched: Date.now(),
+      status: 'loading',
+    };
+  }
+
+  static getDerivedStateFromProps(nextProps, prevState) {
+    const newState = {};
+
+    if(nextProps.eof){
+      newState.status = prevState.status === 'paused' ? 'paused' : 'eof';
+    }
+
+    newState.url = resourceURL(modelFor(nextProps.kind), {
```

Comment:
Would you move this calculation to _beginStreaming since thats the only thing that uses it (`url` is not touched in render). I realize you inherited it, but `url` never belonged in state to begin with.